checks/permissions.go

```diff
@@ -72,41 +78,46 @@ func validatePermission(key string, value interface{}, path string,
 }
 
 func validateMapPermissions(values map[interface{}]interface{}, path string,
-	dl checker.DetailLogger, pdata *permissionCbData) error {
+	dl checker.DetailLogger, pPermissions map[string]bool,
+	ignoredPermissions map[string]bool) error {
 	// Iterate over the permission, verify keys and values are strings.
 	for k, v := range values {
 		key, ok := k.(string)
 		if !ok {
 			//nolint
-			return sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
+			return sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
 		}
 
-		if err := validatePermission(key, v, path, dl, pdata); err != nil {
+		if err := validatePermission(key, v, path, dl, pPermissions, ignoredPermissions); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func recordPermissionWrite(name string, pdata *permissionCbData) {
-	pdata.writePermissions[name] = true
+func recordPermissionWrite(name string, pPermissions map[string]bool) {
+	pPermissions[name] = true
 }
 
-func recordAllPermissionsWrite(pdata *permissionCbData) {
+func recordAllPermissionsWrite(pPermissions map[string]bool) {
 	// Special case: `all` does not correspond
 	// to a GitHub permission.
-	pdata.writePermissions["all"] = true
+	pPermissions["all"] = true
 }
 
-func validateReadPermissions(config map[interface{}]interface{}, path string,
-	dl checker.DetailLogger, pdata *permissionCbData) error {
+func validatePermissions(config map[interface{}]interface{}, path string,
+	dl checker.DetailLogger, pPermissions map[string]bool,
+	ignoredPermissions map[string]bool,
+	acceptUndefined bool) error {
 	var permissions interface{}
 
 	// Check if permissions are set explicitly.
 	permissions, ok := config["permissions"]
 	if !ok {
-		dl.Warn("no permission defined in %v", path)
-		recordAllPermissionsWrite(pdata)
+		if !acceptUndefined {
```

Comment:
would be helpful to add a short comment here to explain why there's this condition. e.g. "Explicit permissions are optional for run level" or something similar. 

Also might be clearer to move the `permissions, ok := config["permissions"]` bits and this check to where they're called from (i.e. validateRunLevelPermissions / validateTopLevelPermissions) and pass the resulting `permissions` to this function instead. This makes the special case (Warn + recordAllPermissionsWrite) closer to where it's needed and easier to follow.

i.e. 

```go
func validateTopLevelPermissions(config map[interface{}]interface{}, path string,
	dl checker.DetailLogger, pdata *permissionCbData) error {
        permissions, ok := config["permissions"]
        // comment explaining why.
        if !ok {
             ... // warn + recordAllPermissionsWrite
            return
        }
	return validatePermissions(permissions, path, dl,
		pdata.topLevelWritePermissions, map[string]bool{},
		false)
}
```

Reply:
good idea. I started adding more and more arguments without realizing it could be simplified. Thanks!